src/ControlSystem/ControlErrors/Shape.hpp

```diff
@@ -119,17 +121,36 @@ struct Shape : tt::ConformsTo<protocols::ControlError> {
     const double lambda_00_coef =
         functions_of_time.at(detail::size_name<Object>())->func(time)[0][0];
 
+    const ylm::Spherepack lambda_spherepack = [](const size_t spectral_size) {
+      // Note: l_max == m_max for the shape map, so the spectral size is
+      // always 2 * (l_max + 1) * (l_max + 1).
+      ASSERT(spectral_size % 2 == 0, "Spectral size "
+                                         << spectral_size << " should be even "
+                                         << " for shape map coefficients.");
+      const auto l_max_plus_one =
+          static_cast<size_t>(sqrt(static_cast<double>(spectral_size) / 2.0));
+      ASSERT(2 * l_max_plus_one * l_max_plus_one == spectral_size,
+             "Shape map spectral size " << spectral_size
+                                        << " cannot be represented by a "
+                                        << "Spherepack with l_max == m_max.");
+      const size_t l_max = l_max_plus_one - 1;
+      return ylm::Spherepack{l_max, l_max};
+    }(lambda_lm_coefs.size());
+
     const auto& ah =
         get<control_system::QueueTags::Horizon<Frame::Distorted, Object>>(
             measurements);
     const auto& ah_coefs = ah.coefficients();
 
-    ASSERT(lambda_lm_coefs.size() == ah_coefs.size(),
-           "Number of coefficients for shape map '"
-               << function_of_time_name << "' (" << lambda_lm_coefs.size()
-               << ") does not match the number of coefficients in the AH "
-                  "Strahlkorper ("
-               << ah_coefs.size() << ").");
+    DataVector resized_ah_coefs{};
```

Comment:
You can do this easier with a nonowning DataVector: set a pointer and size to the original coeffs or assign the resized coeffs